packages/pipeline/src/parsers/copper/index.ts

```diff
@@ -0,0 +1,274 @@
+        entity.websites = 0;
+        entity.phoneNumbers = 0;
+        try {
+            entity.socials = lead.socials.length;
+            entity.websites = lead.websites.length;
+            entity.phoneNumbers = lead.phone_numbers.length;
+        } catch (e) {
+            return entity;
+        }
+
+        return entity;
+    });
+}
+
+/**
+ * Parse response from Copper API /search/activities/
+ *
+ * @param activities - The array of activities returned from the API
+ * @returns Returns an array of Copper Activity entities
+ */
+export function parseActivities(activities: CopperActivityResponse[]): CopperActivity[] {
+    return activities.map(activity => {
+        const entity = new CopperActivity();
+        entity.id = activity.id;
+
+        entity.parentId = activity.parent.id;
+        entity.parentType = activity.parent.type;
+
+        entity.typeId = activity.type.id;
+        entity.typeCategory = activity.type.category.toString();
+        entity.typeName = activity.type.name;
+
+        entity.userId = activity.user_id;
+        entity.dateCreated = activity.date_created * ONE_SECOND;
+        entity.dateModified = activity.date_modified * ONE_SECOND;
+
+        // nullable fields
+        try {
+            entity.oldValueId = activity.old_value.id;
+            entity.oldValueName = activity.old_value.name;
+            entity.newValueId = activity.new_value.id;
+            entity.newValueName = activity.new_value.name;
+        } catch (e) {
+            return entity;
+        }
+        return entity;
+    });
+}
+
+/**
+ * Parse response from Copper API /search/opportunities/
+ *
+ * @param opportunities - The array of opportunities returned from the API
+ * @returns Returns an array of Copper Opportunity entities
+ */
+export function parseOpportunities(opportunities: CopperOpportunityResponse[]): CopperOpportunity[] {
+    return opportunities.map(opp => {
+        const entity = new CopperOpportunity();
+        entity.id = opp.id;
+        entity.name = opp.name;
+        entity.assigneeId = opp.assignee_id;
+        entity.closeDate = opp.close_date;
+        entity.companyId = opp.company_id;
+        entity.companyName = opp.company_name;
+        entity.customerSourceId = opp.customer_source_id;
+        entity.lossReasonId = opp.loss_reason_id;
+        entity.pipelineId = opp.pipeline_id;
+        entity.pipelineStageId = opp.pipeline_stage_id;
+        entity.primaryContactId = opp.primary_contact_id;
+        entity.priority = opp.priority;
+        entity.status = opp.status;
+        entity.tags = opp.tags;
+        entity.interactionCount = opp.interaction_count;
+        entity.monetaryValue = opp.monetary_value;
+        entity.winProbability = opp.win_probability;
+        entity.dateCreated = opp.date_created * ONE_SECOND;
+        entity.dateModified = opp.date_modified * ONE_SECOND;
+        entity.customFields = opp.custom_fields.map(f => f.custom_field_definition_id);
```

Comment:
It looks like this is a one-to-many relationship where an opportunity can have many custom fields? If that's the case, storing custom field ids in a `jsonb` type like this is going to make it harder to do joins. Typically the way that I've set up one-to-many relationships in the past (e.g. where one "parent" has many "children") is to add an `parent_id` field to the child.

Have you considered adding an `opportunity_id` field to the `custom_fields` table instead? It would make joins faster/simpler.

Reply:
I had some uncertainty with this one too. It's actually a many-to-many relationship. We could model it with a junction table, but I'm worried about the sustainability of trying to recreate 
(and maintain!) Copper domain models in our code. Right now it's just an array, which I agree makes querying clunky. 

As a real example, we could be interested in opportunities by integration type (a top level category for custom fields). Two possible solutions:

1. We could change it to a key/value store with `top_level_category -> actual_value`, which means we can do queries like this:
```
SELECT * FROM opportunities GROUP BY custom_fields #>> '{123}' // where 123 is the id for "Integration Type"
```

or,
2. if we wanted to do joins properly, we'd need to add two more tables:
- split the custom field table into one each for the top-level category and the values
- add a junction table with `opportunity_id | custom_field_id | custom_field_parent_id`

This would allow us to do queries like:
```
SELECT COUNT(*) from opportunities as o JOIN custom_fields_relationships as cf 
WHERE o.id = cf.opportunity_id
AND cf.custom_field_parent_id = 123 // the id for the top-level category we are interested in, e.g. "integration type"
GROUP BY cf.custom_field_id
```

My preference is for the first option as it seems simpler and more resilient to Copper domain changes.  What do you think?

Reply:
I implemented option #1 in https://github.com/0xProject/0x-monorepo/pull/1472/commits/cfd7600b766e7c13b6974d5f7af02d61c35073c4 because it was relatively straightforward. Take a look and let me know what you think. :)

Reply:
We talked a bit on Slack. I'm okay with the current solution (option 1).